staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfDecisionHandler(decision int) http.Handler {
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}), longRunningRequestCheck, fakeApfFilter{mockDecision: decision})
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	})
+}
+
+func newApfServer(apfHandler http.Handler) (*httptest.Server, error) {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	}), requestInfoFactory)
+
+	apfServer := httptest.NewServer(handler)
+	return apfServer, nil
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+	assert.NoError(t, err)
+}
+
+func TestApfCanceledRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+			t.Error(err)
+		}
+	}()
+	assert.NoError(t, err)
+
+	//wait metrics to be collected
+	time.Sleep(time.Second)
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_current_inqueue_requests",
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExecutingRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionExecute)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+			t.Error(err)
+		}
+	}()
+	assert.NoError(t, err)
+
+	//wait metrics to be collected
+	time.Sleep(time.Second)
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_current_inqueue_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks",
```

Comment:
apiserver_flowcontrol_read_vs_write_request_count_watermarks appears if and only if apiserver_flowcontrol_read_vs_write_request_count_samples appears.  They are not guaranteed to appear in such a quick test.  OTOH, if the test waits at least a millisecond (see https://github.com/kubernetes/apiserver/blob/09eab1cf2cce55a46e5d9042c90f136d91383e19/pkg/util/flowcontrol/metrics/metrics.go#L125) between `fcmetrics.Register()` and invoking the APF filter, then those two should appear.

Reply:
Thanks for the explanation. Fixed the possible race here.